src/lib/comms/sol-netctl-impl-connman.c

```diff
@@ -1212,3 +1222,274 @@ sol_netctl_service_state_to_str(enum sol_netctl_service_state state)
+
+    context->agent_msg = sd_bus_message_unref(context->agent_msg);
+    context->auth_service = NULL;
+
+    _destory_agent_vector(&context->agent_vector);
+
+    SOL_NULL_CHECK(context->agent, -EINVAL);
+    agent->cancel(data);
+
+    return 0;
+}
+
+static int
+agent_release(sd_bus_message *m, void *data, sd_bus_error *ret_error)
+{
+    struct ctx *context = (struct ctx *)data;
+    const struct sol_netctl_agent *agent = context->agent;
+    void *userdata = (void *)context->agent_data;
+
+    _destory_agent_vector(&context->agent_vector);
+    _release_agent(context);
+
+    SOL_NULL_CHECK(agent, -EINVAL);
+    agent->release(userdata);
+
+    return 0;
+}
+
+static int
+agent_report_error(sd_bus_message *m, void *data, sd_bus_error *ret_error)
+{
+    struct sol_netctl_service *service;
+    struct ctx *context = (struct ctx *)data;
+    const struct sol_netctl_agent *agent = context->agent;
+    const char *path, *err;
+    int r;
+
+    SOL_NULL_CHECK_GOTO(agent, error);
+
+    if (sol_bus_log_callback(m, data, ret_error) < 0)
+        goto error;
+
+    r = sd_bus_message_read(m, "os", &path, &err);
+    SOL_INT_CHECK(r, < 0, r);
+
+    service = find_service_by_path(path);
+    SOL_NULL_CHECK_GOTO(service, error);
+
+    context->agent_msg = sd_bus_message_ref(m);
+    context->auth_service = service;
+    agent->report_error((void *)context->agent_data, service, err);
+
+    return 0;
+error:
+    sd_bus_reply_method_return(m, NULL);
+    return -EINVAL;
+}
+
+static int
+_agent_input_properties(sd_bus_message *m, struct ctx *context)
+{
+    enum sol_netctl_agent_prompt_type type;
+    const struct sol_netctl_agent *agent = context->agent;
+    char *str;
+    struct sol_netctl_agent_input *input;
+    int r;
+
+    r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "{sv}");
+    SOL_INT_CHECK(r, < 0, r);
+
+    _destory_agent_vector(&_ctx.agent_vector);
+    sol_ptr_vector_init(&_ctx.agent_vector);
+
+    do {
+        static const struct sol_str_table table[] = {
+            SOL_STR_TABLE_ITEM("Name",
+                SOL_NETCTL_AGENT_NAME),
+            SOL_STR_TABLE_ITEM("Identity",
+                SOL_NETCTL_AGENT_IDENTITY),
+            SOL_STR_TABLE_ITEM("Passphrase",
+                SOL_NETCTL_AGENT_PASSPHRASE),
+            SOL_STR_TABLE_ITEM("WPS",
+                SOL_NETCTL_AGENT_WPS),
+            { }
+        };
+
+        r = sd_bus_message_enter_container(m, SD_BUS_TYPE_DICT_ENTRY, "sv");
+        SOL_INT_CHECK_GOTO(r, < 1, end);
+
+        str = NULL;
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &str);
+        SOL_INT_CHECK(r, < 0, r);
+
+        if (str)
+            type = sol_str_table_lookup_fallback(table,
+                sol_str_slice_from_str(str),
+                SOL_NETCTL_AGENT_UNKNOWN);
+        else
+            type = SOL_NETCTL_AGENT_UNKNOWN;
+
+        if (type != SOL_NETCTL_AGENT_UNKNOWN) {
+            int is_exist = false;
```

Comment:
rename this variable to 'exists'. I see that you're using it as a bool, so change it from int to bool